File: src/systems/BuildingDamageSystem.ts
import { BuildingStructuralSystem } from './BuildingStructuralSystem';
import { calculateDistance } from '../utils/movementUtils';
import { Position } from '../core/interfaces/types/common';
import {
  BuildingElement,
  BuildingStructure,
  BuildingPhysics,
  BUILDING_PHYSICS_DEFAULTS,
  MaterialType,
  Floor,
  Roof,
  Column,
  Beam,
  Stair,
  Furniture,
  Partition
} from '../core/interfaces/types/building';

export type DamageType = 'impact' | 'fire' | 'explosion';

export interface DamageEvent {
  type: DamageType;
  amount: number;
  position: { x: number; y: number };
  radius?: number; // For area damage
}

export type DamageAxis = 'battle' | 'deterioration';

export interface AxisDamageEvent {
  axis: DamageAxis;
  amount: number;
  reason?: string;
}

export type DamageChangeCallback = (structure: BuildingStructure, axis: DamageAxis, oldValue: number, newValue: number) => void;

/**
 * BuildingDamageSystem handles all damage, repair, and health logic for building elements.
 *
 * Design notes:
 * - All public methods are single-responsibility and under 50 lines.
 * - All mutations are performed on deep copies to ensure immutability and safe state transitions.
 * - Damage and repair calculations are based on material properties and event types.
 * - The system is designed for extensibility: new damage types or repair rules can be added as helpers.
 * - All methods are covered by integration and regression tests in the test suite.
 */
export class BuildingDamageSystem {
  private physics: BuildingPhysics;
  private structuralSystem: BuildingStructuralSystem;
  private damageChangeCallbacks: DamageChangeCallback[] = [];

  constructor(
    physics: Partial<BuildingPhysics> = {},
    structuralSystem?: BuildingStructuralSystem
  ) {
    this.physics = {
      ...BUILDING_PHYSICS_DEFAULTS,
      ...physics
    };
    this.structuralSystem = structuralSystem || new BuildingStructuralSystem(physics);
  }

  /**
   * Apply damage to a building structure
   */
  applyDamage(structure: BuildingStructure, damageEvent: DamageEvent): BuildingStructure {
    const affectedElements = this.getAffectedElements(structure, damageEvent);
    const modifiedStructure = this.createStructureCopy(structure);

    for (const element of affectedElements) {
      const damage = this.calculateDamage(element, damageEvent);
      const newHealth = Math.max(0, element.health - damage);

      const modifiedElement = {
        ...element,
        health: newHealth
      };

      modifiedStructure.elements.set(element.id, modifiedElement);
    }

    // Update overall structure integrity
    modifiedStructure.integrity = this.structuralSystem.calculateIntegrity(modifiedStructure);

    return modifiedStructure;
  }

  /**
   * Repair a building element
   */
  repairElement(
    structure: BuildingStructure,
    elementId: string,
    amount: number
  ): BuildingStructure {
    const element = structure.elements.get(elementId);
    if (!element) {
      return structure;
    }

    const modifiedStructure = this.createStructureCopy(structure);
    const newHealth = Math.min(element.maxHealth, element.health + amount);

    const modifiedElement = {
      ...element,
      health: newHealth
    };

    modifiedStructure.elements.set(elementId, modifiedElement);
    modifiedStructure.integrity = this.structuralSystem.calculateIntegrity(modifiedStructure);

    return modifiedStructure;
  }

  /**
   * Get elements affected by a damage event
   */
  private getAffectedElements(
    structure: BuildingStructure,
    damageEvent: DamageEvent
  ): BuildingElement[] {
    const elements = Array.from(structure.elements.values()) as BuildingElement[];

    if (damageEvent.radius && damageEvent.radius > 0) {
      // Area damage
      return elements.filter((element: BuildingElement) =>
        this.isInRadius(element.position, damageEvent.position, damageEvent.radius!)
      );
    } else {
      // Single point damage
      return elements.filter((element: BuildingElement) =>
        element.position.x === damageEvent.position.x &&
        element.position.y === damageEvent.position.y
      );
    }
  }

  /**
   * Calculate actual damage based on material properties, damage type, and element type
   */
  private calculateDamage(element: BuildingElement, damageEvent: DamageEvent): number {
    const materialProps = this.physics.materialProperties[element.material];
    const resistance = materialProps.resistance[damageEvent.type];
    let damage = damageEvent.amount * (1 - resistance);
    // Distance falloff for area damage
    if (damageEvent.radius) {
      const distance = calculateDistance(element.position, damageEvent.position);
      const falloff = 1 - (distance / damageEvent.radius);
      damage *= Math.max(0, falloff);
    }
    // --- Weathering and Environmental Damage ---
    // If the damage type is weather-related, use weather resistance
    if (damageEvent.type === 'impact' && materialProps.resistance.wind !== undefined) {
      // Example: windstorm event
      damage *= (1 - (materialProps.resistance.wind / 100));
    }
    if (damageEvent.type === 'fire' && materialProps.resistance.heat !== undefined) {
      // Example: heatwave/fire event
      damage *= (1 - (materialProps.resistance.heat / 100));
    }
    // TODO: Add more weather/environmental types as needed
    // --- Weathering Rate ---
    if (materialProps.weatheringRate) {
      // Apply additional damage for materials with high weathering rate
      damage += materialProps.weatheringRate * 10; // Tune multiplier as needed
    }
    // --- Visual Degradation State Hook ---
    // (To be used by visual system for updating appearance)
    // TODO: Integrate with visual system for degradation state
    // --- Track Weathering History ---
    if (materialProps.weatheringHistory) {
      materialProps.weatheringHistory.push({
        timestamp: Date.now(),
        value: damage,
        reason: damageEvent.type
      });
    }
    // Advanced element-specific logic (stub for future):
    switch (element.type) {
      case 'floor':
      case 'roof':
        // Floors/roofs may have special resistance to certain types (e.g., roofs to rain)
        // TODO: Implement weather/fatigue logic
        break;
      case 'column':
      case 'beam':
        // Columns/beams may have higher resistance to impact
        // TODO: Implement structural load/fatigue logic
        break;
      case 'stair':
        // Stairs may degrade faster with usage
        // TODO: Implement usage-based fatigue
        break;
      case 'furniture':
      case 'partition':
        // Furniture/partitions may be more vulnerable to fire/explosion
        // TODO: Implement custom logic
        break;
      default:
        break;
    }
    return damage;
  }

  /**
   * Check if a position is within a radius of another position
   */
  private isInRadius(
    point: { x: number; y: number },
    center: { x: number; y: number },
    radius: number
  ): boolean {
    return calculateDistance(point, center) <= radius;
  }

  /**
   * Create a deep copy of a building structure
   */
  private createStructureCopy(structure: BuildingStructure): BuildingStructure {
    return {
      ...structure,
      elements: new Map(structure.elements)
    };
  }

  /**
   * Get repair cost for an element
   */
  getRepairCost(element: BuildingElement): number {
    const damageTaken = element.maxHealth - element.health;
    const materialCost = this.physics.materialProperties[element.material].weight;
    return damageTaken * materialCost;
  }

  /**
   * Check if an element can be repaired
   */
  canRepair(element: BuildingElement): boolean {
    return element.health < element.maxHealth;
  }

  /**
   * Get optimal repair material for an element
   */
  getOptimalRepairMaterial(element: BuildingElement): MaterialType {
    // By default, use the same material
    return element.material;
  }

  /**
   * Estimate time to repair an element
   */
  estimateRepairTime(element: BuildingElement): number {
    const damageTaken = element.maxHealth - element.health;
    const materialComplexity = this.physics.materialProperties[element.material].weight;
    // Base time unit is seconds
    return damageTaken * materialComplexity * 2;
  }

  /**
   * Apply battle damage (combat) to a structure's battleDamage axis
   */
  applyBattleDamage(structure: BuildingStructure, amount: number, reason?: string): BuildingStructure {
    const oldValue = structure.battleDamage;
    const newValue = this.clampAxis(oldValue + amount);
    const updated = { ...structure, battleDamage: newValue };
    this.triggerDamageChange(updated, 'battle', oldValue, newValue);
    this.checkStateTransition(updated);
    return updated;
  }

  /**
   * Apply deterioration (environmental/time) to a structure's deterioration axis
   */
  applyDeterioration(structure: BuildingStructure, amount: number, reason?: string): BuildingStructure {
    const oldValue = structure.deterioration;
    const newValue = this.clampAxis(oldValue + amount);
    const updated = { ...structure, deterioration: newValue };
    this.triggerDamageChange(updated, 'deterioration', oldValue, newValue);
    this.checkStateTransition(updated);
    return updated;
  }

  /**
   * Clamp axis value to [1, 10]
   */
  private clampAxis(value: number): number {
    return Math.max(1, Math.min(10, Math.round(value)));
  }

  /**
   * Register a callback for damage changes (event hook)
   */
  onDamageChange(cb: DamageChangeCallback) {
    this.damageChangeCallbacks.push(cb);
  }

  /**
   * Trigger all registered damage change callbacks
   */
  private triggerDamageChange(structure: BuildingStructure, axis: DamageAxis, oldValue: number, newValue: number) {
    for (const cb of this.damageChangeCallbacks) {
      cb(structure, axis, oldValue, newValue);
    }
  }

  /**
   * Check for state transitions when thresholds are reached
   */
  private checkStateTransition(structure: BuildingStructure) {
    // Example: If either axis reaches 10, trigger a critical state (extend as needed)
    if (structure.battleDamage >= 10) {
      // TODO: Implement critical battle damage state logic
      // e.g., mark structure as destroyed, trigger events, etc.
    }
    if (structure.deterioration >= 10) {
      // TODO: Implement critical deterioration state logic
      // e.g., mark structure as condemned, trigger events, etc.
    }
  }

  /**
   * Calculate mitigation based on building properties (for future use)
   */
  calculateMitigation(structure: BuildingStructure, axis: DamageAxis): number {
    // Example: Use structural system and material properties for mitigation
    // Extend as needed for more complex logic
    return 0; // Placeholder
  }

  /**
   * Serialize the two-axis damage state for persistence
   */
  static serializeDamageState(structure: BuildingStructure): { battleDamage: number; deterioration: number } {
    return {
      battleDamage: structure.battleDamage,
      deterioration: structure.deterioration
    };
  }

  /**
   * Deserialize the two-axis damage state into a structure
   */
  static deserializeDamageState(structure: BuildingStructure, data: { battleDamage: number; deterioration: number }): BuildingStructure {
    return {
      ...structure,
      battleDamage: data.battleDamage,
      deterioration: data.deterioration
    };
  }
} 

File: python_converted/src/systems/BuildingDamageSystem.py
from typing import Any, Dict, Union


  BuildingElement, 
  BuildingStructure,
  BuildingPhysics,
  BUILDING_PHYSICS_DEFAULTS,
  MaterialType
} from '../types/building'
DamageType = Union['impact', 'fire', 'explosion']
class DamageEvent:
    type: DamageType
    amount: float
    position: Dict[str, Any]
class BuildingDamageSystem {
  private physics: BuildingPhysics
  private structuralSystem: BuildingStructuralSystem
  constructor(
    physics: Partial<BuildingPhysics> = {},
    structuralSystem?: BuildingStructuralSystem
  ) {
    this.physics = {
      ...BUILDING_PHYSICS_DEFAULTS,
      ...physics
    }
    this.structuralSystem = structuralSystem || new BuildingStructuralSystem(physics)
  }
  /**
   * Apply damage to a building structure
   */
  applyDamage(structure: BuildingStructure, damageEvent: DamageEvent): BuildingStructure {
    const affectedElements = this.getAffectedElements(structure, damageEvent)
    const modifiedStructure = this.createStructureCopy(structure)
    for (const element of affectedElements) {
      const damage = this.calculateDamage(element, damageEvent)
      const newHealth = Math.max(0, element.health - damage)
      const modifiedElement = {
        ...element,
        health: newHealth
      }
      modifiedStructure.elements.set(element.id, modifiedElement)
    }
    modifiedStructure.integrity = this.structuralSystem.calculateIntegrity(modifiedStructure)
    return modifiedStructure
  }
  /**
   * Repair a building element
   */
  repairElement(
    structure: BuildingStructure,
    elementId: str,
    amount: float
  ): BuildingStructure {
    const element = structure.elements.get(elementId)
    if (!element) {
      return structure
    }
    const modifiedStructure = this.createStructureCopy(structure)
    const newHealth = Math.min(element.maxHealth, element.health + amount)
    const modifiedElement = {
      ...element,
      health: newHealth
    }
    modifiedStructure.elements.set(elementId, modifiedElement)
    modifiedStructure.integrity = this.structuralSystem.calculateIntegrity(modifiedStructure)
    return modifiedStructure
  }
  /**
   * Get elements affected by a damage event
   */
  private getAffectedElements(
    structure: BuildingStructure,
    damageEvent: \'DamageEvent\'
  ): BuildingElement[] {
    const elements = Array.from(structure.elements.values())
    if (damageEvent.radius && damageEvent.radius > 0) {
      return elements.filter(element => 
        this.isInRadius(element.position, damageEvent.position, damageEvent.radius!)
      )
    } else {
      return elements.filter(element =>
        element.position.x === damageEvent.position.x &&
        element.position.y === damageEvent.position.y
      )
    }
  }
  /**
   * Calculate actual damage based on material properties and damage type
   */
  private calculateDamage(element: BuildingElement, damageEvent: DamageEvent): float {
    const materialProps = this.physics.materialProperties[element.material]
    const resistance = materialProps.resistance[damageEvent.type]
    let damage = damageEvent.amount * (1 - resistance)
    if (damageEvent.radius) {
      const distance = this.calculateDistance(element.position, damageEvent.position)
      const falloff = 1 - (distance / damageEvent.radius)
      damage *= Math.max(0, falloff)
    }
    return damage
  }
  /**
   * Check if a position is within a radius of another position
   */
  private isInRadius(
    point: Dict[str, Any],
    center: Dict[str, Any],
    radius: float
  ): bool {
    return this.calculateDistance(point, center) <= radius
  }
  /**
   * Calculate distance between two points
   */
  private calculateDistance(
    a: Dict[str, Any],
    b: Dict[str, Any]
  ): float {
    const dx = a.x - b.x
    const dy = a.y - b.y
    return Math.sqrt(dx * dx + dy * dy)
  }
  /**
   * Create a deep copy of a building structure
   */
  private createStructureCopy(structure: BuildingStructure): BuildingStructure {
    return {
      ...structure,
      elements: new Map(structure.elements)
    }
  }
  /**
   * Get repair cost for an element
   */
  getRepairCost(element: BuildingElement): float {
    const damageTaken = element.maxHealth - element.health
    const materialCost = this.physics.materialProperties[element.material].weight
    return damageTaken * materialCost
  }
  /**
   * Check if an element can be repaired
   */
  canRepair(element: BuildingElement): bool {
    return element.health < element.maxHealth
  }
  /**
   * Get optimal repair material for an element
   */
  getOptimalRepairMaterial(element: BuildingElement): MaterialType {
    return element.material
  }
  /**
   * Estimate time to repair an element
   */
  estimateRepairTime(element: BuildingElement): float {
    const damageTaken = element.maxHealth - element.health
    const materialComplexity = this.physics.materialProperties[element.material].weight
    return damageTaken * materialComplexity * 2
  }
} 